File: rust/src/manifest.rs
//! The Cairn manifest: the canonical, cross-language description of a cache
//! entry.
//!
//! A manifest records the inputs that were hashed to produce a cache key, the
//! outputs that were captured, and provenance metadata. Both the Rust engine
//! and the Go wrapper serialize manifests with identical field names and an
//! identical canonicalization scheme so the two tools are fully
//! interoperable. The wire format is documented in `docs/FORMAT.md`.

use crate::hash::Hasher;
use crate::json::{self, Json};

/// Current manifest format version. Bump on incompatible changes.
pub const FORMAT_VERSION: u64 = 1;

/// A single hashed input file.
#[derive(Clone, Debug, PartialEq)]
pub struct InputEntry {
    /// Logical path (as supplied on the command line), using forward slashes.
    pub path: String,
    /// Content digest (canonical 16-char hex).
    pub digest: String,
    /// Size in bytes.
    pub size: u64,
}

/// A single captured output file.
#[derive(Clone, Debug, PartialEq)]
pub struct OutputEntry {
    /// Logical path, using forward slashes.
    pub path: String,
    /// Content digest (canonical 16-char hex).
    pub digest: String,
    /// Size in bytes.
    pub size: u64,
}

/// A complete cache manifest.
#[derive(Clone, Debug, PartialEq)]
pub struct Manifest {
    /// Format version.
    pub version: u64,
    /// Tool that produced this manifest, e.g. "cairn-rust" or "cairn-go".
    pub producer: String,
    /// The overall cache key derived from inputs (+ optional command).
    pub key: String,
    /// The command associated with this entry, if any (empty for pure hashing).
    pub command: Vec<String>,
    /// Sorted input entries.
    pub inputs: Vec<InputEntry>,
    /// Sorted output entries.
    pub outputs: Vec<OutputEntry>,
}

impl Manifest {
    /// Compute the canonical cache key from a set of inputs and an optional
    /// command.
    ///
    /// The key is defined as the FNV-1a digest of a canonical byte stream:
    ///
    /// ```text
    /// "cairn-key\x00" version "\x00"
    /// for each input (sorted by path):
    ///     path "\x00" digest "\x00" size "\x00"
    /// "cmd\x00"
    /// for each command arg (in order):
    ///     arg "\x00"
    /// ```
    ///
    /// Inputs MUST be sorted by path before calling. The Go implementation
    /// reproduces this stream byte-for-byte.
    pub fn compute_key(inputs: &[InputEntry], command: &[String]) -> String {
        let mut h = Hasher::new();
        h.update(b"cairn-key\x00");
        h.update(FORMAT_VERSION.to_string().as_bytes());
        h.update(b"\x00");
        for input in inputs {
            h.update(input.path.as_bytes());
            h.update(b"\x00");
            h.update(input.digest.as_bytes());
            h.update(b"\x00");
            h.update(input.size.to_string().as_bytes());
            h.update(b"\x00");
        }
        h.update(b"cmd\x00");
        for arg in command {
            h.update(arg.as_bytes());
            h.update(b"\x00");
        }
        h.finalize_hex()
    }

    /// Serialize to canonical JSON.
    pub fn to_json(&self) -> String {
        let inputs = Json::Array(
            self.inputs
                .iter()
                .map(|i| {
                    json::object(vec![
                        ("path", Json::Str(i.path.clone())),
                        ("digest", Json::Str(i.digest.clone())),
                        ("size", Json::Uint(i.size)),
                    ])
                })
                .collect(),
        );
        let outputs = Json::Array(
            self.outputs
                .iter()
                .map(|o| {
                    json::object(vec![
                        ("path", Json::Str(o.path.clone())),
                        ("digest", Json::Str(o.digest.clone())),
                        ("size", Json::Uint(o.size)),
                    ])
                })
                .collect(),
        );
        let command = Json::Array(self.command.iter().map(|c| Json::Str(c.clone())).collect());
        let doc = json::object(vec![
            ("version", Json::Uint(self.version)),
            ("producer", Json::Str(self.producer.clone())),
            ("key", Json::Str(self.key.clone())),
            ("command", command),
            ("inputs", inputs),
            ("outputs", outputs),
        ]);
        doc.encode()
    }

    /// Parse a manifest from canonical (or any well-formed) JSON.
    pub fn from_json(text: &str) -> Result<Manifest, String> {
        let doc = json::parse(text).map_err(|e| e.to_string())?;
        let version = doc
            .get("version")
            .and_then(Json::as_uint)
            .ok_or("missing 'version'")?;
        let producer = doc
            .get("producer")
            .and_then(Json::as_str)
            .ok_or("missing 'producer'")?
            .to_string();
        let key = doc
            .get("key")
            .and_then(Json::as_str)
            .ok_or("missing 'key'")?
            .to_string();
        let command = doc
            .get("command")
            .and_then(Json::as_array)
            .ok_or("missing 'command'")?
            .iter()
            .map(|v| {
                v.as_str()
                    .map(str::to_string)
                    .ok_or("command arg not a string")
            })
            .collect::<Result<Vec<_>, _>>()?;
        let inputs = parse_entries(doc.get("inputs").ok_or("missing 'inputs'")?)?
            .into_iter()
            .map(|(path, digest, size)| InputEntry { path, digest, size })
            .collect();
        let outputs = parse_entries(doc.get("outputs").ok_or("missing 'outputs'")?)?
            .into_iter()
            .map(|(path, digest, size)| OutputEntry { path, digest, size })
            .collect();
        Ok(Manifest {
            version,
            producer,
            key,
            command,
            inputs,
            outputs,
        })
    }
}

fn parse_entries(value: &Json) -> Result<Vec<(String, String, u64)>, String> {
    let arr = value.as_array().ok_or("expected array of entries")?;

File: go/internal/cache/manifest.go
package cache

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// FormatVersion is the manifest format version, matching the Rust engine.
const FormatVersion uint64 = 1

// Entry is a hashed input or captured output file.
type Entry struct {
	Path   string
	Digest string
	Size   uint64
}

// Manifest is the cross-language cache-entry description.
type Manifest struct {
	Version  uint64
	Producer string
	Key      string
	Command  []string
	Inputs   []Entry
	Outputs  []Entry
}

// ComputeKey derives the canonical cache key from inputs and an optional
// command. Inputs MUST be sorted by path first (SortEntries does this). The
// byte stream is identical to Manifest::compute_key in rust/src/manifest.rs.
func ComputeKey(inputs []Entry, command []string) string {
	h := NewHasher()
	h.Update([]byte("cairn-key\x00"))
	h.Update([]byte(strconv.FormatUint(FormatVersion, 10)))
	h.Update([]byte{0})
	for _, in := range inputs {
		h.Update([]byte(in.Path))
		h.Update([]byte{0})
		h.Update([]byte(in.Digest))
		h.Update([]byte{0})
		h.Update([]byte(strconv.FormatUint(in.Size, 10)))
		h.Update([]byte{0})
	}
	h.Update([]byte("cmd\x00"))
	for _, arg := range command {
		h.Update([]byte(arg))
		h.Update([]byte{0})
	}
	return h.Hex()
}

// SortEntries sorts entries by path in place for deterministic keys.
func SortEntries(entries []Entry) {
	sort.Slice(entries, func(i, j int) bool { return entries[i].Path < entries[j].Path })
}

// NormalizePath converts backslashes to forward slashes for portable manifests.
func NormalizePath(p string) string {
	return strings.ReplaceAll(p, "\\", "/")
}

// ToJSON serializes the manifest to canonical, compact JSON with fields in the
// exact order the Rust engine emits: version, producer, key, command, inputs,
// outputs.
func (m *Manifest) ToJSON() string {
	var b strings.Builder
	b.WriteByte('{')
	writeKey(&b, "version")
	b.WriteString(strconv.FormatUint(m.Version, 10))
	b.WriteByte(',')
	writeKey(&b, "producer")
	writeJSONString(&b, m.Producer)
	b.WriteByte(',')
	writeKey(&b, "key")
	writeJSONString(&b, m.Key)
	b.WriteByte(',')
	writeKey(&b, "command")
	writeStringArray(&b, m.Command)
	b.WriteByte(',')
	writeKey(&b, "inputs")
	writeEntryArray(&b, m.Inputs)
	b.WriteByte(',')
	writeKey(&b, "outputs")
	writeEntryArray(&b, m.Outputs)
	b.WriteByte('}')
	return b.String()
}

func writeKey(b *strings.Builder, key string) {
	writeJSONString(b, key)
	b.WriteByte(':')
}

func writeStringArray(b *strings.Builder, items []string) {
	b.WriteByte('[')
	for i, it := range items {
		if i > 0 {
			b.WriteByte(',')
		}
		writeJSONString(b, it)
	}
	b.WriteByte(']')
}

func writeEntryArray(b *strings.Builder, entries []Entry) {
	b.WriteByte('[')
	for i, e := range entries {
		if i > 0 {
			b.WriteByte(',')
		}
		b.WriteByte('{')
		writeKey(b, "path")
		writeJSONString(b, e.Path)
		b.WriteByte(',')
		writeKey(b, "digest")
		writeJSONString(b, e.Digest)
		b.WriteByte(',')
		writeKey(b, "size")
		b.WriteString(strconv.FormatUint(e.Size, 10))
		b.WriteByte('}')
	}
	b.WriteByte(']')
}

// writeJSONString writes a JSON string literal with the same escaping rules as
// the Rust writer (RFC 8259, lowercase \u for control chars).
